Make /var/lib/kubelet as shared during startup

jsafrane opened the pull request:
This is part of ~~https://github.com/kubernetes/community/pull/589~~ https://github.com/kubernetes/community/pull/659

We'd like kubelet to be able to consume mounts from containers in the future, therefore kubelet should make sure that `/var/lib/kubelet` has shared mount propagation to be able to see these mounts. 

On most distros, root directory is already mounted with shared mount propagation and this code will not do anything. On older distros such as Debian Wheezy, this code detects that `/var/lib/kubelet` is a directory on `/` which has private mount propagation and kubelet bind-mounts `/var/lib/kubelet` as rshared.

Both "regular" linux mounter and `NsenterMounter` are updated here.

@kubernetes/sig-storage-pr-reviews @kubernetes/sig-node-pr-reviews 
@vishh 

Release note:
```release-note
Kubelet re-binds /var/lib/kubelet directory with rshared mount propagation during startup if it is not shared yet.
```

jsafrane commented on cmd/kubelet/app/server.go:
```diff
@@ -861,6 +861,12 @@ func RunKubelet(kubeFlags *options.KubeletFlags, kubeCfg *componentconfig.Kubele
 	}
 
+	// set up /var/lib/kubelet sharing
+	err = kubeDeps.Mounter.MakeShared(kubeCfg.RootDirectory)
```
I hope RunKubelet is the right place to do this startup initialization. Feel free to suggest a better one.

jsafrane replied:
Answering to myself, `setupDataDirs` looks like a better place.

jsafrane commented:
tested with `KUBE_NODE_OS_DISTRIBUTION=gci go run hack/e2e.go -v  --up` and `KUBE_NODE_OS_DISTRIBUTION=debian`. The patch does not do anything on GCI and it bind-mounts /var/lib/kubelet on Debian and makes it rshared.

Surprisingly, something already bind-mounts /var/lib/kubelet as shared on GCI even though it's not necessary, / is already shared.

jsafrane commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

vishh commented on pkg/util/mount/mount.go:
```diff
@@ -55,6 +55,9 @@ type Interface interface {
 	GetDeviceNameFromMount(mountPath, pluginDir string) (string, error)
+	// MakeShared checks that given path is on a mount with 'shared' mount
+	// propagation. If not, it bind-mounts the path as shared.
+	MakeShared(path string) error
```
nit: Call it `MakeRShared` to reflect the fact that it shares recursively?

jsafrane replied:
Renamed

vishh commented on pkg/util/mount/mount_linux.go:
```diff
@@ -425,3 +434,129 @@ func (mounter *SafeFormatAndMount) getDiskFormat(disk string) (string, error) {
+	if err != nil {
+		return []mountInfo{}, err
+	}
+	for {
```
I wish we could make this a generic utility for reading many kernel APIs that return data more than a page in size.

jsafrane replied:
Done (with some refactoring of `listProcMounts` to use the same function).

vishh commented on pkg/util/mount/mount_linux.go:
```diff
@@ -425,3 +434,129 @@ func (mounter *SafeFormatAndMount) getDiskFormat(disk string) (string, error) {
+	}
+
+	glog.V(2).Infof("Bind-mounting %q with shared mount propagation", path)
+	// mount --bind /var/lib/kubelet /var/lib/kubelet
```
Why not use the `syscall` package? Why exec?

vishh replied:
https://golang.org/pkg/syscall/#Mount

jsafrane replied:
I see. Reworked to use `syscall.Mount`.

vishh commented on pkg/util/mount/mount_linux.go:
```diff
@@ -425,3 +434,129 @@ func (mounter *SafeFormatAndMount) getDiskFormat(disk string) (string, error) {
+// doMakeShared is common implementation of MakeShared on Linux. It checks if
+// path is shared and bind-mounts it as shared if needed. mountCmd and mountArgs
+// are expected to contain mount-like command, doMakeShared will add '--bind
+// <path> <path>' and '--make-shared <path>' to mountArgs.
```
nit: make-rshared

jsafrane replied:
fixed

vishh commented on pkg/util/mount/mount_linux_test.go:
```diff
@@ -181,3 +184,105 @@ func TestGetDeviceNameFromMount(t *testing.T) {
+func TestIsSharedSuccess(t *testing.T) {
+	successMountInfo :=
+		`62 0 253:0 / / rw,relatime shared:1 - ext4 /dev/mapper/ssd-root rw,seclabel,data=ordered
+76 62 8:1 / /boot rw,relatime shared:29 - ext4 /dev/sda1 rw,seclabel,data=ordered
```
Can you add a test case with `/var/lib/kubelet` explicitly marked as `shared`?

jsafrane replied:
added a test with `/var/lib/foo` that is  a mount point and is shared.

vishh reviewed (changes requested):
Just a few nits.

jsafrane commented:
@vishh, thanks for the review, I'll update the PR tomorrow.

I played with this PR and with #41683 to make kubelet mount stuff from containers and it looks like /var/lib/kubelet is not enough and https://github.com/kubernetes/kubernetes/pulls/jsafrane is still needed on machines without systemd. See https://github.com/kubernetes/community/issues/648 for details.

ncdc commented:
/unassign
/assign @derekwaynecarr @sjenning

sjenning commented:
Doesn't seem like this would work for nsenter mounter i.e. containerized kubelet.  The container itself will have already bind mounted `/var/lib/kubelet` as private.  nsentering the host and changing the mount propagation after container start isn't going to change the fact that the containerized kubelet has a private `/var/lib/kubelet` right?

sjenning commented on pkg/util/mount/mount_linux.go:
```diff
@@ -425,3 +434,129 @@ func (mounter *SafeFormatAndMount) getDiskFormat(disk string) (string, error) {
+}
+
+type mountInfo struct {
+	root, mountPoint string
```
not seeing where `root` is ever used

vishh replied:
+1.

jsafrane replied:
removed

sjenning commented:
I'm also not seeing the code for "if / is already rshared, do nothing".  Really the rule should be:

```
for dir in /var/lib/kubelet /var/lib /var /; do
  if dir exists in mountpoints
    if dir is private; then
      remount /var/lib/kubelet as rshared
    fi
    return
  fi
done
```
If you get to the end without a remount, it means the /var/lib/kubelet is under / and is reshared already.  You only need to remount if /var/lib/kubelet or one of its parents directories is 1) is a mountpoint and 2) is private.  I think :smile:

jsafrane commented:
>```
>for dir in /var/lib/kubelet /var/lib /var /; do
>  if dir exists in mountpoints
>   if dir is private; then
>      remount /var/lib/kubelet as rshared
>    fi
>    return
>  fi
>done
>```

That's almost what this PR does - `isShared` checks /proc/self/mountinfo from bottom to top and checks that the first directory that's prefix of /var/lib/kubelet (say /var/)  is shared or not.

jsafrane commented:
> Doesn't seem like this would work for nsenter mounter i.e. containerized kubelet. The container itself will have already bind mounted /var/lib/kubelet as private.

Kubelet in a container must already see `/var/lib/kubelet` from the host at least as slave - kubelet mounts things on the host and it **must** see the result in its own container. We do not need to change this. This PR makes /var/lib/kubelet on the host shareable so **pod containers** can export mounts to /var/lib/kubelet on the host so kubelet will see them via slave mount propagation.

sjenning commented:
> That's almost what this PR does - isShared checks /proc/self/mountinfo from bottom to top and checks that the first directory that's prefix of /var/lib/kubelet (say /var/) is shared or not.

Ah ok.  I was missing two things.  In `HasPrefix()`, prefix is the _second_ argument and searching mountinfo from bottom up ensures that you will see `/var/lib` before `/var`, for example.

sjenning commented:
> Kubelet in a container must already see /var/lib/kubelet from the host at least as slave

Are you saying that for old Debian, containerized kubelet returns an error because `/` is `private` and therefore `/var/lib/kubelet` can't be bind mounted at container start time as `rshared` in the kubelet container?

I guess what I'm saying is that if you start a containerized kubelet with `/var/lib/kubelet` bind mounted in from a private `/` on the host (thus it is `private` in the container), nsenter the host and change it to `rshared`, it is still private in the already running kubelet container right?  You'd have to restart the container to see the updated rshared propagation in the kubelet container.  At least that is my understanding.

jsafrane commented:
> searching mountinfo from bottom up ensures that you will see /var/lib before /var, for example.

Mounting is complicated :-). Let there be these 3 mounts created in this order:

/var -> /dev/sda1
/var/lib/ -> /dev/sda2
/var/lib/kubelet -> /dev/sda3

Now, someone comes and mounts a new tmpfs to `/var`. Kubelet will see empty `/var`, even though there is `/var/lib` and `/var/lib/kubelet` in mountinfo - they were shadowed by new `/var` with tmpfs. New mounts are added to bottom of mountinfo, so processing mountinfo from bottom **is** the right thing.

jsafrane commented:
> if you start a containerized kubelet with /var/lib/kubelet bind mounted in from a private / on the host (thus it is private in the container), nsenter the host and change it to rshared, it is still private in the already running kubelet container right? 

Yes, that's right. You can't run containerized kubelet in this situation and this PR won't change it.

You need to add `docker run -v /var/lib/kubelet:/var/lib/kubelet:rslave`  to your container startup script. And this will fail, because docker can't do it if /var/lib/kubelet is not shared or slave on the host. So you need to fix your Wheezy first. Again, this PR does not change it.

I doubt anyone runs containerized kubelet on Wheezy, it's just too painful... The only brave souls I know about that run kubelet/openshift-node in containers are we at Red Hat.

sjenning commented:
Thanks for explaining :+1: @jsafrane 
/lgtm

jsafrane commented:
Wait, I need to fix @vishh remarks first. Self-removing lgtm (I've never thought I would do that)

derekwaynecarr commented:
@jsafrane -- it would not have merged unless you also had it approved (which it was not)

jsafrane commented on pkg/util/mount/mount_linux.go:
```diff
@@ -425,3 +434,129 @@ func (mounter *SafeFormatAndMount) getDiskFormat(disk string) (string, error) {
+	}
+
+	glog.V(2).Infof("Bind-mounting %q with shared mount propagation", path)
+	// mount --bind /var/lib/kubelet /var/lib/kubelet
```
huh? `syscall.Exec` will replace `/usr/bin/kulebet` in current process with `/bin/mount` and it won't ever return. We need `fork()`+`exec()` and that's exactly what `os.Exec` does.

jsafrane commented:
@vishh, I fixed your remarks a while ago, can you please take a look?

kfox1111 commented:
@vishh ping

lucab commented on pkg/util/mount/nsenter_mount.go:
```diff
@@ -229,3 +230,13 @@ func (n *NsenterMounter) absHostPath(command string) string {
+func (n *NsenterMounter) MakeRShared(path string) error {
+	nsenterCmd := nsenterPath
+	nsenterArgs := []string{
+		"--mount=/rootfs/proc/1/ns/mnt",
```
I think you shouldn't hardcode the `/rootfs` prefix here.

jsafrane replied:
It's hardcoded in whole nsenter mounter... I'll make a constant for it, it's used on too many places, but any serious refactoring should happen in a different PR.

lucab commented on pkg/util/mount/nsenter_mount.go:
```diff
@@ -87,9 +87,10 @@ func NewNsenterMounter() *NsenterMounter {
-	hostRootFsPath     = "/rootfs"
-	hostProcMountsPath = "/rootfs/proc/1/mounts"
-	nsenterPath        = "nsenter"
+	hostRootFsPath        = "/rootfs"
```
Thinking about this, is it possible to hide this prefix behind an helper method which does a `stat()` first and returns `/` as a fallback if `/rootfs` doesn't exist (possibly cached)? As `nsenter` does a setns+fchdir+fchroot internally, this will cover containerized, chrooted and unconstrained kubelet the same way.

jsafrane replied:
It's probably possible, but as a separate PR.

jsafrane committed:
Share /var/lib/kubernetes on startup

Kubelet makes sure that /var/lib/kubelet is rshared when it starts.
If not, it bind-mounts it with rshared propagation to containers
that mount volumes to /var/lib/kubelet can benefit from mount propagation.

jsafrane commented:
Now that https://github.com/kubernetes/community/pull/659 is merged, I squashed the commits and rebased them to make the mount propagation usable on Debian Wheezy. This should follow https://github.com/kubernetes/kubernetes/pull/46444.

@vishh, PTAL.

vishh commented on pkg/util/io/consistentread.go:
```diff
@@ -0,0 +1,45 @@
+		if err != nil {
+			return nil, err
+		}
+		if reflect.DeepEqual(oldContent, newContent) {
```
Would https://golang.org/pkg/bytes/#Compare be faster than Deepequal? Not necessary for this PR itself, but something worth considering because on a high churn cluster we want to reduce kubelet resource overhead and latency.

jsafrane replied:
I am not sure if it's faster, but should not be slower. Changed to `bytes.Compare`.

vishh commented:
@jsafrane there are few more open comments

jsafrane commented:
@vishh fixed & tested on Debian Wheezy (with `KUBE_NODE_OS_DISTRIBUTION=debian cluster/kube-up.sh`) - makes /var/lib/kubelet shared and on GCI and Fedora with systemd where nothing happens.

jsafrane commented:
/retest

jsafrane commented:
/retest
#51692

vishh commented on pkg/util/mount/mount_linux.go:
```diff
@@ -502,3 +480,97 @@ func (mounter *SafeFormatAndMount) getDiskFormat(disk string) (string, error) {
+	}
+
+	// mount --make-rshared /var/lib/kubelet
+	if err := syscall.Mount(path, path, "" /*fstype*/, syscall.MS_SHARED|syscall.MS_REC, "" /*data*/); err != nil {
```
Is a remount flag (MS_REMOUNT) required here?

jsafrane replied:
No, remount is not required. `mount --make-rshared` uses only the fags above as implied by the man page and checked with strace.

vishh commented:
I'd like to see a node e2e test that verifies this feature. This may prevent regressions. e2es can land even after code freeze I think.

jsafrane commented:
I have e2e test for whole propagation in my queue, waiting for #46444 to merge. And if it runs on wheezy we can be quite sure that this PR works as expected.

vishh commented:
/lgtm
/approve

jsafrane commented:
/assign @smarterclayton @thockin 
for approval. PTAL.

jsafrane commented:
/retest

smarterclayton commented:
/approve

jsafrane commented:
/approve no-issue
(just checking... :-)

smarterclayton commented:
/approve no-issue
